api.R

```diff
@@ -1,101 +1,238 @@
+    )
   )
-)
 
+  # Load the training data used for this model
+  training_data_md5 <- metadata$dvc_md5_training_data
+  training_data <- read_parquet(
+    file.path(
+      dvc_bucket,
+      substr(training_data_md5, 1, 2),
+      substr(training_data_md5, 3, nchar(training_data_md5))
+    )
+  )
 
-# Load Model -------------------------------------------------------------------
 
-# Load fit and recipe from file
-fit <- lightsnip::lgbm_load(temp_file_fit)
-recipe <- readRDS(temp_file_recipe)
+  # Load Model -----------------------------------------------------------------
 
-# Extract a sample row of predictors to use for the API docs
-predictors <- recipe$var_info %>%
-  filter(role == "predictor") %>%
-  pull(variable)
-ptype_tbl <- training_data %>%
-  filter(meta_pin == "15251030220000") %>%
-  select(all_of(predictors))
-ptype <- vetiver_create_ptype(model = fit, save_prototype = ptype_tbl)
+  # Load fit and recipe from file
+  fit <- lightsnip::lgbm_load(temp_file_fit)
+  recipe <- readRDS(temp_file_recipe)
 
+  # Extract a sample row of data to use for the API docs
+  ptype_tbl <- training_data %>%
+    filter(meta_pin == "15251030220000")
 
-# Create API -------------------------------------------------------------------
+  # If the model recipe is configured to allow it, strip all chars except
+  # for the predictors from the example row
+  if (predictors_only) {
+    predictors <- recipe$var_info %>%
+      filter(role == "predictor") %>%
+      pull(variable)
+    ptype_tbl <- ptype_tbl %>%
+      filter(meta_pin == "15251030220000") %>%
+      select(all_of(predictors))
+  }
 
-# Create model object and populate metadata
-model <- vetiver_model(fit, "LightGBM", save_prototype = ptype)
-model$recipe <- recipe
-model$pv$round_type <- metadata$pv_round_type
-model$pv$round_break <- metadata$pv_round_break[[1]]
-model$pv$round_to_nearest <- metadata$pv_round_to_nearest[[1]]
+  ptype <- vetiver_create_ptype(model = fit, save_prototype = ptype_tbl)
 
-# Start API
-pr() %>%
-  vetiver_api(model) %>%
-  pr_run(
-    host = "0.0.0.0",
-    port = api_port
+
+  # Create API -----------------------------------------------------------------
+
+  # Create model object and populate metadata
+  model <- vetiver_model(fit, "LightGBM", save_prototype = ptype)
+  model$recipe <- recipe
+  model$pv$round_type <- metadata$pv_round_type
+  model$pv$round_break <- metadata$pv_round_break[[1]]
+  model$pv$round_to_nearest <- metadata$pv_round_to_nearest[[1]]
+
+  return(model)
+}
+
+# Filter the valid runs for the run marked as default
+default_run <- valid_runs %>%
+  dplyr::filter(run_id == default_run_id) %>%
+  dplyr::slice_head()
+
+# Retrieve paths and model objects for all endpoints to be deployed
+all_endpoints <- list()
+for (i in seq_len(nrow(valid_runs))) {
+  run <- valid_runs[i, ]
+  model <- get_model_from_run(
+    run$run_id, run$year, run$dvc_bucket, run$predictors_only
   )
+  all_endpoints[[i]] <- list(
```

Comment:
I don't love the way that `valid_runs` is stored as a tibble and `all_endpoints` is stored as a list, but I'm still trying to get a hang of the best way to do these types of list-of-dict operations in R, and these seemed like the most concise approaches to me. Let me know if there's a cleaner way to do all of these iterations!

Reply:
I think it's fine! And technically a tibble _is_ a list 🙃